lib/collection/src/shard/shard_holder.rs

```diff
@@ -64,6 +71,55 @@ impl ShardHolder {
         shard_ops
     }
 
+    /// Create directory to hold the temporary data for shard with `shard_id`
+    pub async fn create_temporary_shard_dir(
+        collection_path: &Path,
+        shard_id: ShardId,
+    ) -> CollectionResult<PathBuf> {
+        let shard_path = collection_path.join(format!("{shard_id}-temp"));
+        tokio::fs::create_dir_all(&shard_path)
+            .await
+            .map_err(|err| CollectionError::ServiceError {
+                error: format!(
+                    "Can't create shard {shard_id} temporary directory. Error: {}",
+                    err
+                ),
+            })?;
+        Ok(shard_path)
+    }
+
+    /// Initiate temporary shard
+    pub async fn initiate_temporary_shard(
```

Comment:
Initialization of the shard is not necessary to be under the write-lock.

In fact, disk operations involved with `LocalShard::build` and `create_temporary_shard_dir` might require quite some time. Let's split initialization into several steps: 

1. Creation of dirs and temporary shard
2. Locking and inserting shard into the holder (should be fast)